javascript/ql/src/semmle/javascript/Promises.qll

```diff
@@ -283,188 +255,143 @@ private module PromiseFlow {
-    ThenStep() { this.getMethodName() = "then" }
-
-    override predicate load(DataFlow::Node pred, DataFlow::Node succ, string prop) {
+  predicate loadStoreStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
+    // promise definition, e.g. `new Promise()`
+    exists(PromiseDefinition promise | succ = promise |
+      // Copy the value of a resolved promise to the value of this promise.
       prop = valueProp() and
-      pred = getReceiver() and
-      succ = getCallback(0).getParameter(0)
+      pred = promise.getResolveParameter().getACall().getArgument(0)
+    )
+    or
+    // promise creation call, e.g. `Promise.resolve`.
+    exists(PromiseCreationCall promise | succ = promise |
+      // Copy the value of a resolved promise to the value of this promise.
+      not promise instanceof PromiseAllCreation and
+      pred = promise.getValue() and
+      prop = valueProp()
       or
+      pred = promise.(PromiseAllCreation).getArrayNode() and
+      prop = valueProp()
+    )
+    or
+    // promise.then()
+    exists(DataFlow::MethodCallNode call | call.getMethodName() = "then" and succ = call |
+      not exists(call.getArgument(1)) and
       prop = errorProp() and
-      pred = getReceiver() and
-      succ = getCallback(1).getParameter(0)
-    }
-
-    override predicate loadStore(DataFlow::Node pred, DataFlow::Node succ, string prop) {
-      not exists(this.getArgument(1)) and
-      prop = errorProp() and
-      pred = getReceiver() and
-      succ = this
+      pred = call.getReceiver()
       or
       // read the value of a resolved/rejected promise that is returned
       (prop = errorProp() or prop = valueProp()) and
-      pred = getCallback([0 .. 1]).getAReturn() and
-      succ = this
-    }
-
-    override predicate store(DataFlow::Node pred, DataFlow::SourceNode succ, string prop) {
-      prop = valueProp() and
-      pred = getCallback([0 .. 1]).getAReturn() and
-      succ = this
-      or
-      prop = errorProp() and
-      pred = getCallback([0 .. 1]).getExceptionalReturn() and
-      succ = this
-    }
-  }
-
-  /**
-   * A flow step describing the data-flow related to the `.catch` method of a promise.
-   */
-  class CatchStep extends PromiseFlowStep, DataFlow::MethodCallNode {
-    CatchStep() { this.getMethodName() = "catch" }
-
-    override predicate load(DataFlow::Node pred, DataFlow::Node succ, string prop) {
-      prop = errorProp() and
-      pred = getReceiver() and
-      succ = getCallback(0).getParameter(0)
-    }
-
-    override predicate loadStore(DataFlow::Node pred, DataFlow::Node succ, string prop) {
+      pred = call.getCallback([0 .. 1]).getAReturn()
+    )
+    or
+    // promise.catch()
+    exists(DataFlow::MethodCallNode call | call.getMethodName() = "catch" and succ = call |
       prop = valueProp() and
-      pred = getReceiver().getALocalSource() and
-      succ = this
+      pred = call.getReceiver().getALocalSource()
```

Comment:
I understand this code was already like that, but why use `getALocalSource` here? It seems like it should just be `call.getReceiver`.

Reply:
If we don't include `.getALocalSource()`, we will miss e.g. [this sink](https://github.com/github/codeql/blob/master/javascript/ql/test/library-tests/Promises/flow.js#L84).  (I ran the test)

I actually don't remember why it is needed.  
But I think its related that `succ` in `storeStep` must be a `SourceNode`.

Reply:
That's really creepy. The path to that sink shouldn't even need this step all (this step's successor is the `catch` call itself, which in that test is a dead end). You're sure it's not just the `nodes/edges` relations that changed?

In any case, it's not a fault introduced by this PR. Happy to move the investigation into an issue if you prefer.

Reply:
> You're sure it's not just the nodes/edges relations that changed?

I thought I had checked that, but apparently I hadn't.  
It was just an `edges` relation that changed (actually a type-tracking-step relation).